chef/cookbooks/barclamp/libraries/barclamp_library.rb

```diff
@@ -417,21 +418,96 @@ def self.size_to_bytes(s)
       end
     end
 
+    class DependsOn
+      class << self
+        def add(dependency)
+          @recipe_depedencies ||= Mash.new
+          @recipe_depedencies.merge!(dependency)
+        end
+
+        def get(recipe)
+          @recipe_depedencies.fetch(recipe, [])
+        end
+      end
+    end
+
     class Config
       class << self
         attr_accessor :node
 
-        def load(group, barclamp, instance = nil)
-          # If no instance is specified, see if this node uses an instance of
-          # this barclamp and use it
+        def last_two_configs(group, barclamp, instance)
+          prev_cfg = load(group, barclamp, instance)
+          instance = infer_instance(group, barclamp, instance)
+          cur_cfg = Chef::Role.load "#{barclamp}-config-#{instance}"
+          return prev_cfg, cur_cfg.default_attributes[barclamp]
+        end
+
+        def loadattr(hash, attrlist)
+          i = 0
+          return hash if attrlist.empty?
+          while hash.respond_to?(:[])
+            unless hash.key?(attrlist[i])
+              Chef::Log.debug("[smart] hash is missing #{attrlist[i]}")
+              return nil
+            end
+            hash = hash[attrlist[i]]
+            i += 1
+            if attrlist.length == i
+              # Chef::Log.debug("[smart] loadattr return #{hash}")
+              return hash
+            end
+          end
+          Chef::Log.debug("[smart] hash does not respond to []")
```

Comment:
this (and the variable name) is misleading. if variable doesn't respond to `[]` it's most probably not a hash.

Reply:
it could also be a single value. this is exactly when the log would trigger. I generally think that this message could be more informative (e.g. "<param name> is not a collection" or sth like this).

Reply:
exactly, it could be a single value or a collection ( i could not think of a word for that)
and that is why the log says the `collection did not respond to ...` because it is a int/string/....